extensions/amp-story/0.1/page-scaling.js

```diff
@@ -0,0 +1,252 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {Services} from '../../../src/services';
+import {dev, user} from '../../../src/log';
+import {dict} from '../../../src/utils/object';
+import {isExperimentOn} from '../../../src/experiments';
+import {childElementsByTag, matches} from '../../../src/dom';
+import {px, setImportantStyles} from '../../../src/style';
+import {renderAsElement} from './simple-template';
+import {throttle} from '../../../src/utils/rate-limit';
+import {toArray} from '../../../src/types';
+
+
+/** @private @const {number} */
+const MIN_LAYER_WIDTH_PX = 380;
+
+
+/** @private @const {number} */
+const MAX_LAYER_WIDTH_PX = 520;
+
+
+/** @private @const {string} */
+const SCALING_APPLIED_CLASSNAME = 'i-amphtml-story-scaled';
+
+
+/** @private @const {!./simple-template.ElementDef} */
+const SIZER_TEMPLATE = {
+  tag: 'div',
+  attrs: dict({'class': 'i-amphtml-story-page-sizer'}),
+};
+
+
+/** @struct @typedef {{factor: number, width: number, height: number}} */
+let DimensionsDef;
+
+
+/** @typedef {{scale: function(!Element):!Promise}} */
+let PageScalingServiceInterface;
+
+
+/**
+ * @param {!Element} sizer
+ * @return {!DimensionsDef}
+ */
+function targetDimensionsFor(sizer) {
+  const {width, height} = sizer./*OK*/getBoundingClientRect();
+
+  const ratio = width / height;
+
+  const targetWidth = Math.min(MAX_LAYER_WIDTH_PX,
+      Math.max(width, Math.max(1, ratio) * MIN_LAYER_WIDTH_PX));
+
+  const targetHeight = (targetWidth / ratio);
+
+  const factor = width / targetWidth;
+
+  return {width: targetWidth, height: targetHeight, factor};
+}
+
+
+/**
+ * @param {!Element} page
+ * @param {!DimensionsDef} dimensions
+ */
+function applyScaling(page, dimensions) {
+  const {width, height, factor} = dimensions;
+  // TODO(alanorozco): Calculate relative layer dimensions for cases where
+  // they won't fill the entire page.
+  const style = {
+    'width': px(width),
+    'height': px(height),
+    'zoom': `${factor * 100}%`,
+    'box-sizing': 'border-box',
+  };
+  scalableChildren(page).forEach(el => {
+    setImportantStyles(el, style);
+  });
+  markScalingApplied(page);
+}
+
+
+/**
+ * @param {!Element} page
+ * @return {!Array<!Element>}
+ */
+function scalableChildren(page) {
+  return toArray(childElementsByTag(page, 'amp-story-grid-layer'));
+}
+
+
+/**
+ * @param {!Element} page
+ * @return {boolean}
+ */
+function isScalingEnabled(page) {
+  return page.getAttribute('scaling') != 'disabled';
```

Comment:
Not super convinced about this API 🤷

Reply:
I think this is okay, at least for now.  Thinking about it more, we definitely want to try it as opt-out behind an experiment first, so that we don't need publishers to republish stories in order to evaluate the feature.

That said, maybe we should brainstorm the explicit values we expect?  This is logically kind of similar to [`object-fit`](https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit), so maybe we can steal their API?

- No scaling can be `scaling="none"`
- This implementation is like `scaling="contain"` (or, at least for now, omitting the attribute altogether)
- We can add other modes later if we choose to
- Maybe we should name it accordingly?  `content-fit`?

Not really sure about any of the details here, but I think we should check that the attribute is some concrete value instead of checking that it's not disabled.

Reply:
Actually, maybe we can do both opt-in and opt-out? Assuming the attribute names from above (though I'm not particularly set on those names), we can do:

|        | Experiment OFF           | Experiment ON  |
| -------------:|:----------:|:-----:|
| (no attribute) | scaling disabled | scaling enabled |
| `scaling="none"` | scaling disabled | scaling disabled |
| `scaling="contain"` | scaling enabled | scaling enabled |

In other words, explicitly setting a value is always respected, but the default (no attribute) case is controlled by the experiment.  This will allow us to meet two goals:

1. Test the feature on ALL existing stories, without having to change their markup
2. Allow publishers to publish new stories making use of the feature, by explicitly opting in on a story-by-story basis

WDYT?  Is this feasible?  Can we move the experiment check to `isScalingEnabled`?